tket/src/Transformations/PhasedXFrontier.cpp

```diff
@@ -0,0 +1,432 @@
+  }
+  return std::nullopt;
+}
+
+OptEdgeVec PhasedXFrontier::get_all_beta_edges() const {
+  OptEdgeVec beta_edges;
+  std::vector<Vertex> beta_vertices;
+  for (unsigned i = 0; i < circ_.n_qubits(); ++i) {
+    OptEdge beta_e = get_beta_edge(i);
+    beta_edges.push_back(beta_e);
+    if (beta_e) {
+      beta_vertices.push_back(circ_.target(*beta_e));
+    }
+  }
+
+  // for OpType::NPhasedX, we need to reset some betas to zero if they
+  // are shadowed (i.e. if one NPhasedX is behind another one)
+  std::map<Vertex, unsigned> arities = count(beta_vertices);
+  for (auto& e : beta_edges) {
+    if (e) {
+      Vertex v = circ_.target(*e);
+      Op_ptr op = circ_.get_Op_ptr_from_Vertex(v);
+      if (arities[v] != op->n_qubits()) {
+        e = std::nullopt;
+      }
+    }
+  }
+  return beta_edges;
+}
+
+OptVertexVec PhasedXFrontier::get_all_beta_vertices() const {
+  OptVertexVec vertices;
+  for (const auto& e : get_all_beta_edges()) {
+    if (e) {
+      Vertex v = circ_.target(*e);
+      vertices.push_back(v);
+    } else {
+      OptVertex optv;
+      vertices.push_back(optv);
+    }
+  }
+  return vertices;
+}
+
+std::vector<Expr> PhasedXFrontier::get_all_betas() const {
+  std::vector<Expr> betas;
+  for (const auto& v : get_all_beta_vertices()) {
+    Expr beta = 0.;
+    if (v) {
+      Op_ptr op = circ_.get_Op_ptr_from_Vertex(*v);
+      beta = op->get_params().front();
+    }
+    betas.push_back(beta);
+  }
+  return betas;
+}
+
+void PhasedXFrontier::next_interval(unsigned i) {
+  auto& [start, end] = intervals_[i];
+
+  // new interval
+  start = get_interval_start(end);
+  end = get_interval_end(start);
+}
+
+void PhasedXFrontier::next_multiqb(const Vertex& v) {
+  std::set<unsigned> curr_qubits = qubits_ending_in(v);
+  // move forward
+  for (unsigned i : curr_qubits) {
+    next_interval(i);
+  }
+}
+
+Edge PhasedXFrontier::get_interval_end(Edge e) const {
+  Vertex v = circ_.target(e);
+  while (!is_interval_boundary(v)) {
+    auto p = circ_.get_next_pair(v, e);
+    v = p.first;
+    e = p.second;
+  }
+  return e;
+}
+
+Edge PhasedXFrontier::get_interval_start(Edge e) const {
+  Vertex v = circ_.target(e);
+  if (!circ_.detect_final_Op(v)) {
+    e = circ_.get_next_edge(v, e);
+  }
+  return e;
+}
+
+bool PhasedXFrontier::is_interval_boundary(Vertex v) const {
+  Op_ptr op = circ_.get_Op_ptr_from_Vertex(v);
+  OpType optype = op->get_type();
+  if (optype == OpType::NPhasedX) {
+    return false;
+  }
+  return !is_single_qubit_type(optype) || !is_gate_type(optype);
+}
+
+/**
+ * @brief Implements the AbstractSquasher interface for squashing to PhasedX+Rz
+ */
+class PhasedXSquasher : public StandardSquasher {
+ public:
+  PhasedXSquasher()
+      : StandardSquasher(
+            OpTypeSet{OpType::Rz, OpType::PhasedX},
+            CircPool::tk1_to_PhasedXRz) {}
+
+  // Accept any single-qubit gate that has TK1 angles to squash it.
+  bool accepts(Gate_ptr gp) const override {
+    try {
```

Comment:
Don't like this change -- seems like a hack and is less efficient. Maybe we need to add `&& !is_projective_type()`?

Reply:
Mmmm, yes, I agree that this is less efficient. However, just adding predicates until it works doesn't sound right either. Ideally, we would have access to the set of optypes that implement `get_tk1_angles`. Shall I add that?

Reply:
Yes I think that would be useful: `is_single_qubit_unitary_type()`?

Reply:
Perfect, I'll add that.

Reply:
Done. I have kept the parameter as `Gate_ptr` as I will need that in the future anyways.